.clj-kondo/config.edn

```diff
@@ -122,7 +123,108 @@
   {:exclude
    [colorize.core]}
 
-  :unsorted-required-namespaces {:level :warning}
+  :metabase/ns-module-checker
+  {:level :warning
+
+   ;; Map of module name => the "core" external public-facing API namespace(s).
+   ;;
+   ;; `nil` or an empty set means this module does not (yet) have an external public-facing API namespace. Go create
+   ;; one. Otherwise this should be a set of namespace symbols. All namespaces in other modules will only be allowed
+   ;; to use namespaces from this set.
```

Comment:
If I understand correctly, all `metabase.x` prefixes are considered to be modules, and whether they are missing from this map or have a `nil` entry, we will calculate their `module-api-namespaces` as the empty set. This empty set is treated as the set of all possible child namespaces within this module. The nil entries seem to have been added here to organically drive us to define more restrictive endpoints but I don't see a mechanism to ensure that we remember to add entries for new namespaces.

At the offsite we discussed having some convention for the single entry point (e.g. `metabase.upload`) or core (e.g. `metabase.lib.core`). This still seems sensible to me, and due perhaps most to familiarity I prefer the former one. I believe it's more common and we can always override it for the other cases.

The approach taken in the earlier PR was for each prefix to opt-in as being a module, an one factor in this choice was the observation that we have some nested modules e.g. `metabase.sync.sync-metadata` and `metabase.sync.analyze` which are more like grouped top-level modules than sub-modules. As we improve our module boundaries I think that both the ability to group "public" modules within deeper prefixes and to work with submodules will be valuable, but I guess we can put that off for now. I do like how aggressive the "everything in metabase is a module" approach will drive our initial detangling.

### Suggestion:

If we're going for this aggressive approach, why not make the "parent entry point" convention the default and apply that to all modules not found in this map. Since it'll take a bunch of refactoring to get there we could instead use a sentinel value like `:*` instead for this behavior. This will highlight the semantics, and their undesirability, more clearly in the config, and remove a potential footgun of treating the value like an empty collection in the linter. The burn-down of encapsulating each linter will involve just deleting the entry, and new modules will default to enforcing our single-entry convention unless an override is added here. I guess we would rename this key to `api-namespace-overrides`.

In the long run I'm not convinced that we'll need multiple entry points for a given module, that feels like a poor mans submodules + visibility controls.

Reply:
Yeah `:*` is a good idea. 

Multiple entry points is short-term thing. Step 1 is to try to get everything added to the list and then step 2 is cleaning stuff up

Reply:
Ended up adding `:any` as a new sentinel value here, otherwise the default is `#{metabase.my-module}`